### DISCUSSION TITLE
[Feature Request]: Support the public_html web root

### DISCUSSION THREAD

**Discussion by Brugman** [beyondcode/herd-community#794] [Ideas]
I mostly run WordPress sites, a bunch of vanilla PHP sites, and some Laravel sites. WordPress sites and vanilla `index.php` apps get picked up by Herd/Valet automatically if I put them in the root of a project folder.

Like Laravel uses the `public` subfolder to separate the web root from the other files in the project, I and many others have adopted the `public_html` folder a long time ago to match Apache. When you've got this folder committed to git, and used in your deployment procedures, it is not something you want to change on a whim.

Besides putting the site in the project root, Herd/Valet works out of the box with WP and vanilla sites in the `public` folder, but not the `public_html` folder. Could this be added to the core functionality? An even better implementation would let the user configure what folders are checked, become some may prefer `htdocs` or `www` subfolders.

**_You should just add your own local Valet driver!_**
That is easier said than done. I've spent an entire evening playing with modified versions of the [default drivers](https://github.com/laravel/valet/tree/master/cli/Valet/Drivers). The only one that managed to pass all my tests was a slightly modified BasicValetDriver, but that contained none of the features that the WordPress specific driver has, so I know it is subpar. Sure, I could ask or wait for an expert to create a good local Valet driver for every specific CMS that defines a custom subfolder, but considering how generic the request is, I think it would be better solved at a core Herd/Valet level.

I would hesitate to ask this from the Valet project, as its focus is primarily on Laravel, but Herd's audience is much wider, (optimistically) bringing in lots of PHP devs that don't use Laravel. I think it would massively benefit Herd to add this flexibility to its Valet implementation.

**Comment by Brugman** [comment]
I've tried adding my own local Valet driver, again. And it didn't work.

I want to use the default `WordPressValetDriver`, so I copied that to `MyWordPressValetDriver`.
It extends `BasicValetDriver`. That driver looks equivalent to `BasicWithPublicValetDriver`, except the latter makes /public/ work. So I copied `BasicWithPublicValetDriver` to `MyBasicWithPublicValetDriver`.

Now I can modify `MyBasicWithPublicValetDriver` to contain /public_html/, and modify `MyWordPressValetDriver` to extend `MyBasicWithPublicValetDriver` instead. Problem solved? Nope.

It seems I cannot extend a custom driver with another custom driver. Maybe that's an alternative solution. If you could allow that in Herd/Valet, then my previous request which requires UI work could be scratched, but I obviously would prefer a built-in solution with a UI.

**Reply by InforPsico** [reply]
I'm asking the same thing.

I read that they managed to fix it here, but I can't get it to work, I don't know if you've been able to

https://github.com/beyondcode/herd-community/discussions/1207

**Reply by Brugman** [reply]
I have not. I'm giving it another shot, but I have low expectations.

My solution in the mean time has been to symlink public to public_html. This way Herd/Valet sees WP in public but I actually have it stored in public_html.

From what I've read, you and marcamos want a custom subfolder for _Laravel_ sites. The Laravel config already has a reference to `/public` so modifying that seems easier and is not the same thing. I'm looking to merge two configs or extend a custom config with a custom config.

**Comment by Brugman** [comment]
I think I managed to build a working driver for WordPress sites in a public_html subfolder. If anyone wants a different subfolder it should be easy to modify.

https://github.com/Brugman/laravel-herd-valet-drivers/blob/master/MyWordPressPublicHTMLValetDriver.php

I'll leave this FR open because it still has value to be integrated in core.
